sql: fix LOG of very large numbers

madelynnblue opened the pull request:
Previously a LOG of a large number would result in the server consuming
all CPU and hanging. This is because our LOG implementation would take 1/x
at a precision low enough that the division of the large x would result
in a 0. The algorithm would proceed to double it until it reached 0.5,
which obviously would never succeed. Fix this by using a high enough
precision to prevent 0 from being returned.

During the investigation of this bug, it was also discovered that if a
large number was near the vicinity of producing a 0, it would incorrectly
truncate some precision of the division, thus producing slightly incorrect
results to the LOG function.

This thus fixes various bugs and hangs in the POW, LOG, and LN builtins.

Fixes current RSG test failures.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10221)

<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/util/decimal/decimal.go, line 290 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10221#-KUzf49mdpw75E1I289k:-KUzf49mdpw75E1I289l:b-r3l5di) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1453ac0f1919c58c2a290a3d235647f8753577f/pkg/util/decimal/decimal.go#L290)):*
> ```Go
> 		// 2 because one length of the string will cover the 0s after the decimal,
> 		// and the other length of the string covers the digits of the division.
> 		// TODO(mjibson): figure out a way to do this without converting to a string. I
> ```

I think something like `encoding.numDigits` is what you were looking for. It would be nice to avoid the String formatting every time here, but it's your call if you want to look into exposing `numDigits`-like functionality (which is pretty specialized to the encoding process right now) up to here.

I would just ask that we run some of the benchmarks related to this change (and post the results as part of the commit) to make sure the overhead isnt egregious.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10221#-:-KUzfaIzrGLiiElHmoou:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/decimal/decimal.go, line 290 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10221#-KUzf49mdpw75E1I289k:-KV1FZp0tEe04OInO7a2:b-x8uol0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1453ac0f1919c58c2a290a3d235647f8753577f/pkg/util/decimal/decimal.go#L290)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I think something like `encoding.numDigits` is what you were looking for. It would be nice to avoid the String formatting every time here, but it's your call if you want to look into exposing `numDigits`-like functionality (which is pretty specialized to the encoding process right now) up to here.
> 
> I would just ask that we run some of the benchmarks related to this change (and post the results as part of the commit) to make sure the overhead isnt egregious.
</details>
I added benchmarks to the commit message: no change. Usingc numDigits works, but I needed to move it to the decimal package to avoid a cyclic dependency error. It makes more sense to be here anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10221)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> thanks, even better.

---

Reviewed 1 of 2 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/util/decimal/table.go, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10221#-KV1Mq4W4BbHBgOh6120:-KV1Mq4XFnUJg5dtJHaj:bjegw3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd60277b2fdf5664068240a852435653e8713468/pkg/util/decimal/table.go#L30)):*
> ```Go
> // 10 up to this magnitude. For instance, if the tableSize if 3, then the lookup
> // table will store power of 10 values from 10^-3 to 10^3.
> const tableSize = 32
> ```

This name should be made more specific now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10221#-:-KV1N7WJGcA22VBX6dM5:b-kdame0)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: fix LOG of very large numbers

Previously a LOG of a large number would result in the server consuming
all CPU and hanging. This is because our LOG implementation would take 1/x
at a precision low enough that the division of the large x would result
in a 0. The algorithm would proceed to double it until it reached 0.5,
which obviously would never succeed. Fix this by using a high enough
precision to prevent 0 from being returned.

During the investigation of this bug, it was also discovered that if a
large number was near the vicinity of producing a 0, it would incorrectly
truncate some precision of the division, thus producing slightly incorrect
results to the LOG function.

This thus fixes various bugs and hangs in the POW, LOG, and LN builtins.

Fixes current RSG test failures.

Benchmark shows no change:

name          old time/op  new time/op  delta
DecimalLog-4   106µs ± 3%   105µs ± 1%  -1.26%  (p=0.029 n=10+10)
DecimalExp-4  18.6ms ±17%  18.4ms ± 8%    ~     (p=0.897 n=10+8)
DecimalPow-4  1.25ms ±18%  1.25ms ±21%    ~     (p=0.853 n=10+10)
